src/dbnode/persist/fs/wide/entry_checksum_mismatch_checker.go

```diff
@@ -0,0 +1,284 @@
+				// mismatch.
+				if !c.blockReader.Next() {
+					c.exhausted = true
+				} else {
+					batch = c.blockReader.Current()
+					markerIdx = len(batch.Checksums) - 1
+					c.batchIdx = 0
+				}
+
+				return c.entryMismatches(entry), nil
+			} else if compare > 0 {
+				// This is a mismatch on primary that appears before the
+				// marker element. Return mismatch but do not advance iter.
+				return c.entryMismatches(entry), nil
+			}
+
+			// The current batch here is exceeded. Emit the current batch marker as
+			// a mismatch on primary, and advance the block iter.
+			c.indexMismatches(checksum)
+			if !c.blockReader.Next() {
+				// If no further values, add the current entry as a mismatch and return.
+				c.exhausted = true
+				return c.entryMismatches(entry), nil
+			}
+
+			batch = c.blockReader.Current()
+			markerIdx = len(batch.Checksums) - 1
+			c.batchIdx = 0
+			continue
+		}
+
+		if checksum == entry.idChecksum {
+			// Matches: increment batch index and return any gathered mismatches.
+			c.batchIdx = c.batchIdx + 1
+			return c.mismatches, nil
+		}
+
+		for nextBatchIdx := c.batchIdx + 1; nextBatchIdx < markerIdx; nextBatchIdx++ {
+			// NB: read next hashes, checking for index checksum matches.
+			nextChecksum := batch.Checksums[nextBatchIdx]
+			if entry.idChecksum != nextChecksum {
+				continue
+			}
+
+			// Checksum match. Add previous checksums as mismatches.
+			mismatches := c.indexMismatches(batch.Checksums[c.batchIdx:nextBatchIdx]...)
+			c.batchIdx = nextBatchIdx + 1
+			return mismatches, nil
+		}
+
+		checksum = batch.Checksums[markerIdx]
+		// NB: this is the last element in the batch. Check ID against MARKER.
+		if entry.idChecksum == checksum {
+			if compare != 0 {
+				// Checksums match but IDs do not. Treat as invariant violation.
+				return nil, c.invariant(batch.EndMarker, checksum, entry)
+			}
+
+			c.indexMismatches(batch.Checksums[c.batchIdx:markerIdx]...)
+			// ID and checksum match. Advance the block iter and return empty.
+			if !c.blockReader.Next() {
+				c.exhausted = true
+			} else {
+				batch = c.blockReader.Current()
+				markerIdx = len(batch.Checksums) - 1
+				c.batchIdx = 0
+			}
+
+			return c.mismatches, nil
+		}
+
+		// Checksums do not match.
+		if compare > 0 {
+			// This is a mismatch on primary that appears before the
+			// marker element. Return mismatch but do not advance iter.
+			return c.entryMismatches(entry), nil
+		}
+
+		// Current value is past the end of this batch. Mark all in batch as
+		// mismatches, and receive next batch.
+		c.indexMismatches(batch.Checksums[c.batchIdx:]...)
+		if !c.blockReader.Next() {
+			// If no further values, add the current entry as a mismatch and return.
+			c.exhausted = true
+			return c.entryMismatches(entry), nil
+		}
+
+		batch = c.blockReader.Current()
+		markerIdx = len(batch.Checksums) - 1
+		c.batchIdx = 0
+	}
+}
+
+func (c *entryChecksumMismatchChecker) Drain() []ReadMismatch {
```

Comment:
Looks `Drain` this is only used for testing, guess it should be private then.

Reply:
Yeah, this will be used to drain any remaining values in the wide index checksum list as checksum mismatches 👍